test/powershell/Modules/Microsoft.PowerShell.Core/Enter-PSHostProcess.Tests.ps1

```diff
@@ -2,54 +2,158 @@
-
-        if ($IsWindows) {
-            $powershell_started = New-TemporaryFile
-            $si.FileName = "powershell"
-            $si.Arguments = "-noexit -command 'powershell' >'$powershell_started'"
-            $powershell = [System.Diagnostics.Process]::Start($si)
+    Context "By Process Id" {
+        BeforeAll {
+            $pwsh_started = New-TemporaryFile
+            $si = [System.Diagnostics.ProcessStartInfo]::new()
+            $si.FileName = "pwsh"
+            $si.Arguments = "-noexit -command 'pwsh' > '$pwsh_started'"
+            $si.RedirectStandardInput = $true
+            $si.RedirectStandardOutput = $true
+            $si.RedirectStandardError = $true
+            $pwsh = [System.Diagnostics.Process]::Start($si)
+
+            if ($IsWindows) {
+                $powershell_started = New-TemporaryFile
+                $si.FileName = "powershell"
+                $si.Arguments = "-noexit -command 'powershell' >'$powershell_started'"
+                $powershell = [System.Diagnostics.Process]::Start($si)
+            }
+
         }
 
-    }
+        AfterAll {
+            $pwsh | Stop-Process
+            Remove-Item $pwsh_started -Force -ErrorAction SilentlyContinue
+
+            if ($IsWindows) {
+                $powershell | Stop-Process
+                Remove-Item $powershell_started -Force -ErrorAction SilentlyContinue
+            }
+        }
 
-    AfterAll {
-        $pwsh | Stop-Process
-        Remove-Item $pwsh_started -Force -ErrorAction SilentlyContinue
+        It "Can enter and exit another PSHost" {
+            Wait-UntilTrue { (Get-PSHostProcessInfo -Id $pwsh.Id) -ne $null }
 
-        if ($IsWindows) {
-            $powershell | Stop-Process
-            Remove-Item $powershell_started -Force -ErrorAction SilentlyContinue
+            "Enter-PSHostProcess -Id $($pwsh.Id) -ErrorAction Stop
+            `$pid
+            Exit-PSHostProcess" | pwsh -c - | Should -Be $pwsh.Id
         }
-    }
 
-    It "Can enter and exit another PSHost" {
-        Wait-UntilTrue { Test-Path $pwsh_started }
+        It "Can enter and exit another Windows PowerShell PSHost" -Skip:(!$IsWindows) {
+            Wait-UntilTrue { (Get-PSHostProcessInfo -Id $powershell.Id) -ne $null }
+
+            "Enter-PSHostProcess -Id $($powershell.Id) -ErrorAction Stop
+            `$pid
+            Exit-PSHostProcess" | pwsh -c - | Should -Be $powershell.Id
+        }
 
-        "enter-pshostprocess -id $($pwsh.Id)`n`$pid`nexit-pshostprocess" | pwsh -c - | Should -Be $pwsh.Id
+        It "Can enter using NamedPipeConnectionInfo" {
+            $npInfo = [System.Management.Automation.Runspaces.NamedPipeConnectionInfo]::new($pwsh.Id)
+            $rs = [runspacefactory]::CreateRunspace($npInfo)
+            $rs.Open()
+            $ps = [powershell]::Create()
+            $ps.Runspace = $rs
+            $ps.AddScript('$pid').Invoke() | Should -Be $pwsh.Id
+            $rs.Dispose()
```

Comment:
Please put this in `finally` block (try {} finally {})

Reply:
fixed.